src/xtgeo/common/sys.py

```diff
@@ -99,126 +143,126 @@ def generic_hash(gid, hashmethod="md5"):
         KeyError: String in hashmethod has an invalid option
 
     .. versionadded:: 2.14
+
     """
-    validmethods = {
+    validmethods: dict[str, Callable] = {
         "md5": hashlib.md5,
         "sha256": hashlib.sha256,
         "blake2b": hashlib.blake2b,
     }
 
-    mhash = None
-    if isinstance(hashmethod, str):
+    if isinstance(hashmethod, str) and hashmethod in validmethods:
         mhash = validmethods[hashmethod]()
     elif isinstance(hashmethod, BuiltinFunctionType):
         mhash = hashmethod()
+    else:
+        raise ValueError(f"Invalid hash method provided: {hashmethod}")
 
     mhash.update(gid.encode())
     return mhash.hexdigest()
 
 
-class _XTGeoFile(object):
-    """A private class for file/stream handling in/out of XTGeo and CXTGeo.
+class _XTGeoFile:
+    """
+    A private class for file/stream handling in/out of XTGeo and CXTGeo.
 
     Interesting attributes:
 
-    xfile = _XTGeoFile(..some Path or  str or BytesIO ...)
+    xfile = _XTGeoFile(..some Path or str or BytesIO ...)
 
     xfile.name: The absolute path to the file (str)
     xfile.file: The pathlib.Path instance
     xfile.memstream: Is True if memory stream
 
-    xfile.exist(): Returns True (provided mode 'r') if file exists, always True for 'w'
+    xfile.exists(): Returns True (provided mode 'r') if file exists, always True for 'w'
     xfile.check_file(...): As above but may raise an Excpetion
     xfile.check_folder(...): For folder; may raise an Excpetion
     xfile.splitext(): return file's stem and extension
     xfile.get_cfhandle(): Get C SWIG file handle
     xfile.cfclose(): Close current C SWIG filehandle
 
-
     """
 
-    def __init__(self, fobj, mode="rb", obj=None):
-        self._file = None  # Path instance or BytesIO memory stream
-        self._tmpfile = None
+    def __init__(
+        self,
+        filelike: Union[str, pathlib.Path, io.BytesIO, io.StringIO],
+        mode: Literal["rb", "wb"] = "rb",
```

Comment:
This list is very incomplete as far as write modes go -- but there is only special handling for these two (and their first character). Not clear to me if we should include all write mode permutations or not, but since this is internal I guess it's less important to sort that out now.

Reply:
Yes we can sort that later